cmd/internal/labeler/labeler.go

```diff
@@ -455,8 +379,95 @@ func createNumaNodeMappingLabel(mapping map[int][]string) string {
 	return strings.Join(parts, "_")
 }
 
-func (l *labeler) printLabels() {
+func (l *labeler) printLabelsToFile(labelFile string) error {
+	f, err := os.OpenFile(labelFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return fmt.Errorf("failed to open file (%s): %w", labelFile, err)
+	}
+
+	defer f.Close()
+
+	for key, val := range l.labels {
+		if _, err := f.WriteString(key + "=" + val + "\n"); err != nil {
+			return fmt.Errorf("failed to write label (%s=%s) to file: %w", key, val, err)
+		}
+	}
+
+	return nil
+}
+
+func CreateAndPrintLabels(sysfsDRMDir string) {
+	l := newLabeler(sysfsDRMDir)
+
+	if err := l.createLabels(); err != nil {
+		klog.Warningf("failed to create labels: %+v", err)
+
+		return
+	}
+
 	for key, val := range l.labels {
 		fmt.Println(key + "=" + val)
 	}
 }
+
+// Gathers node's GPU labels on channel trigger or timeout, and write them to a file.
+// The created label file is deleted on exit (process dying).
+func Run(sysfsDrmDir, nfdFeatureFile string, updateInterval time.Duration, scanResources chan bool) {
+	l := newLabeler(sysfsDrmDir)
+
+	interruptChan := make(chan os.Signal, 1)
+	signal.Notify(interruptChan, syscall.SIGTERM, syscall.SIGINT, syscall.SIGHUP, syscall.SIGQUIT)
+
+	klog.V(1).Info("Starting GPU labeler")
+
+	running := true
+
+	for running {
+		timeout := time.After(updateInterval)
+
+		select {
+		case <-timeout:
+		case <-scanResources:
+		case interrupt := <-interruptChan:
+			klog.V(2).Infof("Interrupt %d received", interrupt)
+
+			running = false
+
+			continue
+		}
+
+		klog.V(1).Info("Ext resources scanning")
+
+		err := l.createLabels()
+		if err != nil {
+			klog.Warningf("label creation failed: %+v", err)
+
+			continue
+		}
+
+		if l.labelsChanged {
+			klog.V(1).Info("Writing labels")
+
+			if err := l.printLabelsToFile(nfdFeatureFile); err != nil {
```

Comment:
To make sure NFD got the whole file, not just a partial write of it, I think this needs do atomic file rename, e.g. from a subdirectory not processed by NFD.

(This was not a problem with binary hooks because those were run by NFD itself, which catches their whole output.)

Reply:
Although possible, I find it unlikely and not critical. Mainly because if NFD reads the file while it's being written, the output would be partial, but it would be completed on the next iteration.

To be able to write it somewhere and then move it, we'd have to mount a temp directory to the Pod as its file system is readonly.

Reply:
I was looking at the NFD documentation. It also notes to make the writing atomic, so I guess I'll implement that.

Reply:
Implemented. Although, using the nfd's features.d directory is somewhat risky operation. NFD could travel through the subdirectories. But at least currently it doesn't.